UploadDaemon/UploadDaemon.cs

```diff
@@ -0,0 +1,168 @@
+                config = Config.ReadFromCentralConfigFile();
+            }
+            catch (Exception e)
+            {
+                logger.Error(e, "Failed to read config file {configPath}", Config.ConfigFilePath);
+                Environment.Exit(1);
+                return;
+            }
+
+            logger.Info("Starting upload daemon v{uploaderVersion}", Assembly.GetExecutingAssembly().GetName().Version.ToString());
+            var uploader = new UploadDaemon(config);
+            uploader.RunOnce();
+            uploader.ScheduleRegularUploads();
+            uploader.WaitForNotifications();
+        }
+
+        private static bool IsAlreadyRunning()
+        {
+            Process current = Process.GetCurrentProcess();
+            // If this process is not running from the exe file, we allow it
+            // Not sure when this happens, though
+            if (Assembly.GetExecutingAssembly().Location.Replace("/", "\\") != current.MainModule.FileName)
+            {
+                return false;
+            }
+
+            return Process.GetProcessesByName(current.ProcessName).Where(process => process.Id != current.Id).Any();
+        }
+
+        public UploadDaemon(Config config)
+        {
+            fileSystem = new FileSystem();
+
+            this.config = config;
+            if (config.DisableSslValidation)
+            {
+                HttpClientUtils.DisableSslValidation();
+            }
+
+            uploadTask = new UploadTask(config, fileSystem, new UploadFactory());
+        }
+
+        /// <summary>
+        /// Runs the timer action once synchronously.
+        /// </summary>
+        public void RunOnce()
+        {
+            lock(SequentialUploadsLock)
+            {
+                uploadTask.Run();
+            }
+        }
+
+        /// <summary>
+        /// Schedule uploads on a regular intervall.
+        /// </summary>
+        private void ScheduleRegularUploads()
+        {
+            Timer timer = new Timer();
+            timer.Elapsed += (sender, args) => RunOnce();
+            timer.Interval = TimerIntervalInMilliseconds;
+            timer.Enabled = true;
+        }
+
+        /// <summary>
+        /// Waits for notifications from subsequent executions of the Daemon. Terminates on
+        /// interrupt from the console's cancel key.
+        /// </summary>
+        private void WaitForNotifications()
+        {
+            while (true) // wait for indefinitely many commands
+            {
+                using (var pipeServerStream = new NamedPipeServerStream(DaemonControlPipeName, PipeDirection.In, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous))
+                {
+                    pipeServerStream.WaitForConnection();
+                    using (var streamReader = new StreamReader(pipeServerStream))
+                    {
+                        var command = streamReader.ReadLine();
```

Comment:
do we have to trim or strip whitespace (i.e. the trailing \r\n) here?

Reply:
Nope, [`ReadLine()` does that for us](https://docs.microsoft.com/en-us/dotnet/api/system.io.streamreader.readline?redirectedfrom=MSDN&view=netframework-4.7.2#remarks).